trin-core/src/ethereum/bridge.rs

```diff
@@ -0,0 +1,227 @@
+use async_trait::async_trait;
+use ethers_core::types::{
+    Address, Block, Bytes, EIP1186ProofResponse, Transaction, TransactionReceipt, TxHash, H256,
+    U256, U64,
+};
+use ethers_providers::{
+    Ipc, JsonRpcClient, Middleware, Provider, ProviderError, PubsubClient, SubscriptionStream, Ws,
+};
+use log::error;
+use url::Url;
+
+use std::error::Error;
+use std::fmt;
+use std::path::Path;
+
+// TODO: Add more information to client and serialization errors.
+/// An error that can occur in communications with a bridge node.
+#[derive(Debug)]
+pub enum BridgeError {
+    EthClientError,
+    SerializationError,
+    ConnectionFailure,
+}
+
+impl fmt::Display for BridgeError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::EthClientError => write!(f, "bridge ETH client error"),
+            Self::SerializationError => write!(f, "bridge serialization error"),
+            Self::ConnectionFailure => write!(f, "bridge connection failure"),
+        }
+    }
+}
+
+impl Error for BridgeError {}
+
+impl From<serde_json::Error> for BridgeError {
+    fn from(_: serde_json::Error) -> Self {
+        BridgeError::SerializationError
+    }
+}
+
+impl From<ProviderError> for BridgeError {
+    fn from(error: ProviderError) -> Self {
+        match error {
+            ProviderError::JsonRpcClientError(_) => BridgeError::EthClientError,
+            ProviderError::SerdeJson(_) | ProviderError::HexError(_) => {
+                BridgeError::SerializationError
+            }
+            _ => BridgeError::ConnectionFailure,
+        }
+    }
+}
+
+/// A type alias for a `Result` whose `Err` variant holds a `BridgeError`.
+type Result<T> = std::result::Result<T, BridgeError>;
+
+/// An Ethereum bridge client.
+#[async_trait]
+pub trait EthBridge {
```

Comment:
I feel like I've seen you explain the reasoning somewhere else, but I can't remember where... But, why did you choose the `EthBridge` / `GethBridge` names? Something like `HistoryBridge` and `StateBridge` seem much more intuitive to me, especially to help delineate the various responsibilities of each type, as this code gets used elsewhere in the codebase.

Reply:
Yeah, so I felt that `EthBridge` and `GethBridge` more accurately captured why there are two separate structures in the first place, denoting the additional set of capabilities of `geth`. I guess I saw that these names being more general was an advantage instead of a disadvantage. I cannot find a mention of "bridge nodes" in the docs on the header/transaction gossip network, but if those networks would also require a bridge, then there is not going to be some naming confusion around state versus history versus gossip. Instead, these bridges are more general components.

Reply:
Cool, ok hmmm I guess one question is that with this architecture, am I correct in assuming that you expect there to be a single bridge node process regardless of how many overlay networks are activated?

Reply:
That's correct. One thing to note with this implementation is that if there are multiple overlay networks that wish to subscribe to new block headers, then the bridge will initiate and manage multiple subscriptions (one for each overlay) instead of a single subscription with a fan-out pattern. This is definitely not ideal, but I'm not sure that the performance improvement would be worth the additional complexity at this time.

My thought was that this is more of a "minimum viable bridge" and that there is room for this optimization along with others when we want to build a more optimized implementation.